ports/raspberrypi/common-hal/busio/UART.c

```diff
@@ -24,238 +24,135 @@
  * THE SOFTWARE.
  */
 
-#include "shared-bindings/microcontroller/__init__.h"
 #include "shared-bindings/busio/UART.h"
 
-#include "mpconfigport.h"
-#include "lib/utils/interrupt_char.h"
-#include "py/gc.h"
+#include "py/stream.h"
 #include "py/mperrno.h"
 #include "py/runtime.h"
-#include "py/stream.h"
-#include "supervisor/shared/translate.h"
 #include "supervisor/shared/tick.h"
+#include "lib/utils/interrupt_char.h"
 
-#define UART_DEBUG(...) (void)0
-// #define UART_DEBUG(...) mp_printf(&mp_plat_print __VA_OPT__(,) __VA_ARGS__)
-
-// Do-nothing callback needed so that usart_async code will enable rx interrupts.
-// See comment below re usart_async_register_callback()
-// static void usart_async_rxc_callback(const struct usart_async_descriptor *const descr) {
-//     // Nothing needs to be done by us.
-// }
+#include "src/rp2_common/hardware_gpio/include/hardware/gpio.h"
 
 #define NO_PIN 0xff
 
+#define UART_INST(uart) (((uart) ? uart1 : uart0))
+
+#define TX  0
+#define RX  1
+#define CTS 2
+#define RTS 3
+
+typedef enum {
+    STATUS_FREE = 0,
+    STATUS_IN_USE,
+    STATUS_NEVER_RESET
+} uart_status_t;
+
+static uart_status_t uart_status[2];
+
+void uart_reset(void) {
+    for (uint8_t num = 0; num < 2; num++) {
+        if (uart_status[num] == STATUS_IN_USE) {
+            uart_status[num] = STATUS_FREE;
+            uart_deinit(UART_INST(num));
+        }
+    }
+}
+
+void never_reset_uart(uint8_t num) {
+    uart_status[num] = STATUS_NEVER_RESET;
+}
+
+static uint8_t get_free_uart() {
+    uint8_t num;
+    for (num = 0; num < 2; num++) {
+        if (uart_status[num] == STATUS_FREE) {
+            break;
+        }
+        if (num) {
+            mp_raise_RuntimeError(translate("All UART peripherals in use"));
```

Comment:
To match existing error text:

```suggestion
            mp_raise_RuntimeError(translate("All UART peripherals are in use"));
```